CleverRaven/Cataclysm-DDA
# Tiring dodges

#### Summary
None 

#### Purpose of change
Continuation of #47576

#### Describe the solution

- [x] Check for potential damage before deciding to dodge
- [x] Check for probability of dodge before trying to dodge
- [x] Move check_dodge out of monattack and  into Creature
- [x] Unify all the scattered `get_dodge` to use `dodge_check` every time an actual dodge is supposed to take place
- [x] Make use of `dodge_roll()`
- [x] Move ` dodges_left--` from `on_dodge` to `on_try_dodge`
- [x] Make sure `dodge_check` is called where needed
- [x] doudble check there's no free dodge left-overs
- [x] Make sure tests are still working
- [x] Projectile attacks

Also consume dodge_attempt inside `on_dodge_try` instead of on succesfull dodge which will probably reduce effective chance to dodge since you'll only get one attempt per turn instead of one succes per turn

#### Describe alternatives you've considered



#### Testing

- Get naked
- Get high dodge
- Get invincible
- Spawn a slime
- Wait and dodge
- See stamina go down
- When stamina get around two bars, the dodges don't happen anymore and the stamina doesn't get lower

Check that dodging projectile cares about stamina
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/41293484/f3168ff8-b23c-41f5-a59d-e2e3a0a7cdd7)

Check that you get a message about not trying to doge special attacks
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/41293484/e6ef9e86-d461-4233-a1c3-3c721fe83170)


#### Additional context

We still need to unify dodges, this makes two parallele way to doge:
- Special attacks are dodged via `dodge_check` which only tries to dodge if we have 80% or more chance of succes (should make it 50%?) , and only if the attack would do non 0 damage. Alternatively if the attacks were to do more than half the current hp of the target part in damage it tries to doge no matter the chances
- Regular attack go through deal_melee_attack which always calls `on_try_dodge`

I didn't manage to unify everything because deal_melee_attack sometimes use flat hit roll value, and sometimes random value from dice rolls and sometimes random values from rng_normal and I have no idea how to unify those
